# Проектная работа
# Тема: "SQL" Анализ базы данных сервиса для чтения книг

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задания**
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Задачи проекта**
- Описать цели исследования;
- Исследовать таблицы — вывести первые строки;
- Сделать по одному SQL-запросу для решения каждого задания;
- Вывести результат каждого запроса в тетрадке;
- Описать выводы по каждой из решённых задач.

**Описание данных**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![Описание изображения](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2)

<h1>План работы<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Исследование-таблиц" data-toc-modified-id="Исследование-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование таблиц</a></span></li><li><span><a href="#Решение-задач" data-toc-modified-id="Решение-задач-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Решение задач</a></span><ul class="toc-item"><li><span><a href="#Задача-№1" data-toc-modified-id="Задача-№1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Задача №1</a></span></li><li><span><a href="#Задача-№2" data-toc-modified-id="Задача-№2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Задача №2</a></span></li><li><span><a href="#Задача-№3" data-toc-modified-id="Задача-№3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Задача №3</a></span></li><li><span><a href="#Задача-№4" data-toc-modified-id="Задача-№4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Задача №4</a></span></li><li><span><a href="#Задача-№5" data-toc-modified-id="Задача-№5-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Задача №5</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Рекомендации" data-toc-modified-id="Рекомендации-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Рекомендации</a></span></li></ul></div>

## Загрузка данных

In [6]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

pd.set_option('display.max_columns', 30)

In [7]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Исследование таблиц

Выведем первые 5 строк каждой таблицы:

In [11]:
#таблица books
books = '''SELECT * FROM books LIMIT 5'''

con = engine.connect()
pd.io.sql.read_sql(sql=text(books), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [12]:
#таблица authors
authors = '''SELECT * FROM authors LIMIT 5'''

pd.io.sql.read_sql(sql=text(authors), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [13]:
#таблица publishers
publishers = '''SELECT * FROM publishers LIMIT 5'''

pd.io.sql.read_sql(sql=text(publishers), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [14]:
#таблица ratings
ratings = '''SELECT * FROM ratings LIMIT 5'''

pd.io.sql.read_sql(sql=text(ratings), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [15]:
#таблица reviews
reviews = '''SELECT * FROM reviews LIMIT 5'''

pd.io.sql.read_sql(sql=text(reviews), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Решение задач

### Задача №1

Посчитайте, сколько книг вышло после 1 января 2000 года

In [19]:
query_1 = '''
SELECT COUNT(*) AS book_count
FROM books
WHERE publication_date > '2000-01-01';
'''

pd.io.sql.read_sql(sql=text(query_1), con = con)

book_count
0         819

**Вывод:**

После 1 января 2000 года вышло 819 книг.

### Задача №2

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [23]:
query_2 = '''
SELECT b.book_id, 
       b.title,
       COUNT(DISTINCT r.review_id) AS review_count,
       AVG(rt.rating) AS average_rating
FROM books AS b
LEFT JOIN reviews AS r ON b.book_id = r.book_id
LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
GROUP BY b.book_id
ORDER BY review_count DESC, average_rating DESC;
'''

pd.io.sql.read_sql(sql=text(query_2), con=con)

book_id                                              title  review_count  \
0        948                            Twilight (Twilight  #1)             7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
2        299  Harry Potter and the Chamber of Secrets (Harry...             6   
3        656                                     The Book Thief             6   
4        734                                   The Glass Castle             6   
..       ...                                                ...           ...   
995      191  Disney's Beauty and the Beast (A Little Golden...             0   
996      387                               Leonardo's Notebooks             0   
997      221                          Essential Tales and Poems             0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999      808                            The Natural Way to Draw             0   

     average_rating  
0          3.662500  
1          4.414634  
2          4.287500  
3          4.264151  
4          4.206897  
..              ...  
995        4.000000  
996        4.000000  
997        4.000000  
998        3.666667  
999        3.000000  

[1000 rows x 4 columns]

**Вывод:**

Распределение книг по среднему рейтингу:
- количество книг со средним рейтингом 5 баллов - 44,
- со средним рейтингом от 4 баллов до 4.9 - 520,
- со средним рейтингом от 3 баллов до 3.9 - 401,
- со средним рейтингом от 2 баллов до 2.9 - 34,
- со средним рейтингом от 0 баллов до 1.9 - 1,

Распределение книг по количеству отзывов:
- 7 отзывов - наибольшее количество отзывов у книги Twilight (Twilight #1), средний рейтинг - 3.6
- 6 отзывов - 15 книг, с разными рейтингами, среди них, например, Water for Elephants (рейтинг 3.9), The Glass Castle (рейтинг 4.2), Harry Potter and the Prisoner of Azkaban (рейтинг 4.4), The Curious Incident of the Dog in the Night-Time (рейтинг 4.4)
- 5 отзывов - 60 книг
- 4 отзыва - 160 книг
- 3 отзыва - 277 книг
- 2 отзыва - 444 книги
- 1 отзыв - 37 книг
- 0 отзывов - 6 книг

Также есть книги без отзывов, но с рейтингом.

### Задача №3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [27]:
query_3 = '''
SELECT p.publisher_id,
       p.publisher,
       COUNT(b.book_id) AS book_count
FROM publishers AS p
JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id, p.publisher
ORDER BY book_count DESC
LIMIT 1;
'''

pd.io.sql.read_sql(sql=text(query_3), con = con)

publisher_id      publisher  book_count
0           212  Penguin Books          42

**Вывод:**

Наибольшее количество книг толще 50 страниц выпустило издательство Penguin Books - оно выпустило 42 такие книги.

### Задача №4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [31]:
query_4 = '''
SELECT a.author_id,
       a.author,
       AVG(bb.average_rating) AS avg_rating
FROM books AS b
JOIN (
    SELECT b.book_id, 
           b.title,
           COUNT(rt.rating) AS count_rating,
           AVG(rt.rating) AS average_rating
    FROM books AS b
    LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
    GROUP BY b.book_id
    HAVING COUNT(rt.rating) >= 50
    ORDER BY average_rating DESC
) AS bb ON b.book_id = bb.book_id
JOIN authors AS a ON b.author_id = a.author_id
GROUP BY a.author_id
ORDER BY avg_rating DESC
LIMIT 1;
'''

pd.io.sql.read_sql(sql=text(query_4), con=con)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.283844

**Вывод:**

Автор с самой высокой средней оценкой книг (учитывая только книги с 50 и более оценками) - J.K. Rowling. Средний рейтинг её книг составляет 4.28.

### Задача №5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [35]:
query_5 = '''
SELECT AVG(review_count) AS average_review_count
FROM (
  SELECT username,
         COUNT(review_id) AS review_count
  FROM reviews
  WHERE username IN (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 48
  )
  GROUP BY username
) AS subquery;
'''

pd.io.sql.read_sql(sql=text(query_5), con = con)

average_review_count
0                  24.0

**Вывод:**

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, - 24.

## Общий вывод

1) После 1 января 2000 года вышло 819 книг.

2) В базе данных содержится 1000 книг

Распределение книг по среднему рейтингу:
- количество книг со средним рейтингом 5 баллов - 44,
- со средним рейтингом от 4 баллов до 4.9 - 520,
- со средним рейтингом от 3 баллов до 3.9 - 401,
- со средним рейтингом от 2 баллов до 2.9 - 34,
- со средним рейтингом от 0 баллов до 1.9 - 1,

Распределение книг по количеству отзывов:
- 7 отзывов - наибольшее количество отзывов у книги Twilight (Twilight #1), средний рейтинг - 3.6
- 6 отзывов - 15 книг, с разными рейтингами, среди них, например, Water for Elephants (рейтинг 3.9), The Glass Castle (рейтинг 4.2), Harry Potter and the Prisoner of Azkaban (рейтинг 4.4), The Curious Incident of the Dog in the Night-Time (рейтинг 4.4)
- 5 отзывов - 60 книг
- 4 отзыва - 160 книг
- 3 отзыва - 277 книг
- 2 отзыва - 444 книги
- 1 отзыв - 37 книг
- 0 отзывов - 6 книг

Также есть книги без отзывов, но с рейтингом.

3) Наибольшее количество книг толще 50 страниц выпустило издательство Penguin Books - оно выпустило 42 такие книги.

4) Автор с самой высокой средней оценкой книг (учитывая только книги с 50 и более оценками) - J.K. Rowling. Средний рейтинг её книг составляет 4.28.

5) Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, - 24.

## Рекомендации

- Исходя из анализа данных, можно сделать вывод, что пользователи предпочитают книги с высоким рейтингом. Рекомендуется уделить внимание и продвижению книг с рейтингом 4 и выше, так как они имеют наибольшее количество отзывов и высокую оценку. 

- Стоит обратить внимание на книги с небольшим количеством отзывов или без отзывов, но с хорошим рейтингом, и предложить пользователям ознакомиться с ними.

- Издательство Penguin Books выпустило наибольшее количество книг, которые превышают 50 страниц. Рекомендуется рассмотреть возможность установления партнерства с такими издательствами для обеспечения разнообразия и качества предлагаемых книг подписчикам сервиса.

- Пользователи, которые оставляют больше 48 оценок, имеют среднее количество обзоров 24. Рекомендуется разработать механизмы стимулирования активности пользователей, например, предоставление бонусов, рейтинговых систем или возможности получения дополнительных преимуществ, чтобы поощрить пользователей оставлять больше обзоров и рейтингов.

- Популярный автор: J.K. Rowling имеет самую высокую среднюю оценку книг. Рекомендуется выделить работы этого автора и активно продвигать его книги в сервисе.

Учитывая выводы из проведенного анализа, эти рекомендации помогут определить стратегию развития сервиса чтения книг по подписке, привлечь больше пользователей и удовлетворить их интересы и предпочтения.